### ISSUE TITLE
CMake for 1.4.0 does not conform to CMake's FindFLTK

### ISSUE THREAD

**Issue by MatthiasWM** [fltk/fltk#311]
The CMake maintainers were nice enough to prove a FindFLTK script in their standard distribution. However FLTK 1.4.0 is no longer compatible to it. Docs are here: https://cmake.org/cmake/help/latest/module/FindFLTK.html .

Basically, `sudo cmake --install .` should install FLTK on any platform, so that `find_library(FLTK REQUIRED)` should set up all the variables mentioned in the CMake docs (and probably more to make life easy for our users). 

If installing FLTK is not an option, `add_subdirectory(my_fltk_clone)` should also "[just work](https://www.fltk.org/newsgroups.php?s0+gfltk.issues+v1727+T)". 

Is anyone on the team capable of fixing up our current, very elaborate, CMake system? Or should I try to get some outside help in?

**Comment by MoAlyousef** [comment]
Currently `add_subdirectory(my_fltk_clone)` works. 

It's still not naive friendly though:
1- It requires adding the fltk binary dir to the target_include_directrories for the FL/abi-version.H header:
```cmake
target_include_directories(main PRIVATE my_fltk_clone ${CMAKE_BINARY_DIR}/my_fltk_clone})
```

2- Naively one would expect run cmake -Bbin to only build the library. It currently also builds the whole suite of tests, you need either
```cmake
set(FLTK_BUILD_TEST OFF CACHE BOOL " " FORCE)
```
or pass it in the cmake command line or gui.

**Comment by Albrecht-S** [comment]
@MatthiasWM I know of the incompatibilities and I intend to improve our CMake files before the release of FLTK 1.4.

However, there's not much we can do about the outdated FindFLTK module provided by CMake. Even if we tried to "fix" this by pushing our own to the CMake folks this would -- obviously -- only mitigate this for users with the latest CMake version (or at least the one that has this updated FindFLTK module).

The best solution would be to provide a FindFLTK module with the FLTK sources and that's what I'm striving to do. I did already work on it but couldn't finish it.

I believe that this is also the best solution for our minimal CMake requirements. We always try not to require the "best and latest" version of our dependencies, and CMake is such a dependency.

This is also more or less the solution suggested by the CMake folks who noticed that maintaining FindXXX modules for all software is a nightmare.

Our current README.CMake.txt file clearly states that the CMake FindFLTK version is not compatible and IMHO this is not going to change but I'll try to find a better solution.

**Comment by engelsman** [comment]
CMake's FindFLTK.html document is seriously out-of-date. It mentions FLTK2!

See https://www.fltk.org/articles.php?L834 "Using CMake to build and FLTK application"
which I originally wrote after struggling to get anything to work based on that howto.
Albrecht has some specific guidelines in README.CMake.txt which gets you further.

As far as I can tell, CMake's FindFLTK module was a user contribution that has basically
withered on the vine, with nobody responsible for following it up and providing updates
and it's a disincentive to people trying to use CMake to build FLTK if the "official" info
doesn't work.

**Comment by MatthiasWM** [comment]
Ah, great to see that this is a know issue. Yes, the FindFLTK script is more of a hassle than a help (there is a FIndFLTK2 as well...), but it's the solution that Forum users seem to find first :-( . Maybe we can teach CMake to honor the FindFLTK docs (and legacy apps), but also provide better CMake properties?!

Yes, I read the README.cmake, but installing for Xcode fails (it works via command line), and for VisualStudio, installation is not mentioned (it somehow works via Developer Shell in Administrator mode. But starting a new CMake project with `find_library(FLTK)` does not seem to work without setting per-platform paths.

I looked in CMake for hours and I found the install repo which remembers which library was installed and where. FLTKConfig.cmake (IIRC), which will be installed together with FLTK, could hopefully be made compatible with the dreaded FindFLTK for all platforms?

I'll be happy to help to make this perfect. Even for me, it is not obvious and a hassle to start a new project that uses Fluid and links with FLTK. IMHO it should be "naive friendly" (great expression, @MoAlyousef).

**Comment by Albrecht-S** [comment]
I'll try to do my best. We really need some improvements in CMake and I'm considering to raise the minimum CMake requirements from 3.2.3 to a higher version if needed. More "modern CMake" support will likely require higher CMake versions as well but I'd still like to stay conservative.

OTOH: CMake can be installed pretty easily on all major platforms. Linux distros may be a little behind, hence I believe we shouldn't force the user to install anything (CMake) from outside the distro's package sources. It's difficult.

And, as you found out, the CMake FindFLTK module (even one created by us) would require lots of backwards compatibility hooks. I can tell you I tried hard, but then I ran out of time and gave up. But I will definitely attack this again!

**Comment by Albrecht-S** [comment]
Current status: with the advent of "Modern CMake" many of the issues mentioned in this thread are gone, and all the variables mentioned in FindFLTK.cmake are obsolete. I believe that meanwhile FLTK's CMake support is "naive friendly".

@MoAlyousef wrote:
> 1- It requires adding the fltk binary dir to the target_include_directrories for the FL/abi-version.H header ...

The required include directories are now provided by the FLTK targets, for instance 'fltk::fltk' and/or 'fltk::images'. Users don't need to use `target_include_directrories()` to link to FLTK libraries anymore.

> 2- Naively one would expect run cmake -Bbin to only build the library. It currently also builds the whole suite of tests ...

Currently (in Git) this is no longer true if FLTK detects that it is built as a subproject (in a subdirectory of the main build). In this case the test programs are no longer built. However, if FLTK is built in the main (root) CMakeLists.txt, then building the test programs in the `test/` folder is still enabled by default. Should we change this?

Meanwhile we also have options to disable building `fluid` and `fltk-options` whose defaults are `ON` for both. `fluid` can and will often be needed by projects using FLTK, hence I left it ON in all cases. I'm not sure that `fltk-options` will be needed or wanted by users building FLTK but it's really small and fast to build.

Building the (X)Forms compatibility library and the OpenGL support lib can be disabled as well, but both are also ON by default.

That said, the only remaining question is how to deal with FindFLTK.cmake. I thought of writing one to be bundled and installed with FLTK so users can use it in a "modern" version. But it's not as straight-forward as one might think. The CMake CONFIG module installed with FLTK "knows" how it was built and which libraries and include directories it needs (i.e. which ones users must include/link to). An independent FindFLTK.cmake module can never know how a particular FLTK library has been built. The only "solution" I can imagine would be to let FindFLTK.cmake use fltk-config to query the libraries and include directories the user would need. Since fltk-config will always be present this is theoretically doable, but it sounds weird and not desirable - and error-prone as well.

After thinking longer about compatibility of FindFLTK.cmake I came to the conclusion that it would be best to declare it's obsolete and should not be used. I can try to request that it should be removed from the CMake distributions because it's no longer helpful. I don't know...

**Comment by MoAlyousef** [comment]
Hi @Albrecht-S 
I've tested the CMake modifications with a C++ project and it works. I'm also glad that I don't need to add gdiplus in a conditional for target_link_libraries on Windows.
Regarding FindFLTK.cmake, I think CMake itself now advocates the use of `<lib>Config.cmake` instead. 
Regarding fluid, I'm glad building it is optional. Even though I like it and have Rust bindings for it, building it by default felt unnecessary, and actually caused some build failures in fltk-rs in some situitations. 
As for tests, I guess that's reasonable. I'm not sure whether CMake's fetchContent would treat it as a subdir or not, I'll have to check that out.

**Comment by Albrecht-S** [comment]
> I'm not sure whether CMake's fetchContent would treat it as a subdir or not, I'll have to check that out.

FetchContent builds the "fetched" project in a subdirectory.

I was just writing this when I saw your confirmation (edit) above. Thanks.

**Comment by Albrecht-S** [comment]
> Regarding FindFLTK.cmake, I think CMake itself now advocates the use of <lib>Config.cmake instead.

Yes, ISTR that I also read this somewhere. Thanks for your confirmation. I would even go for not writing one before the release of FLTK 1.4.0 and then wait for feedback. If users need one we can still do it - or clarify how they should work with CMake's CONFIG mode...
